test/persistence-hooks.suite.js

```diff
@@ -274,26 +274,26 @@ module.exports = function(dataSource, should) {
           });
       });
 
-      // TODO(bajtos) Enable this test for all connectors that
-      // provide optimized implementation of findOrCreate.
-      // The unoptimized implementation does not trigger the hook
-      // when an existing model was found.
-      it.skip('triggers `before save` hook when found', function(done) {
-        TestModel.observe('before save', pushContextAndNext());
+      if (dataSource.connector.findOrCreate) {
+        it('triggers `before save` hook when found', function(done) {
+          TestModel.observe('before save', pushContextAndNext());
 
-        TestModel.findOrCreate(
-          { where: { name: existingInstance.name } },
-          { name: existingInstance.name },
-          function(err, record, created) {
-            if (err) return done(err);
-            observedContexts.should.eql(aTestModelCtx({ instance: {
-              id: record.id,
-              name: existingInstance.name,
-              extra: undefined
-            }}));
-            done();
-          });
-      });
+          var id = uid();
+
+          TestModel.findOrCreate(
+            { where: { name: existingInstance.name } },
+            { id: id, name: existingInstance.name },
```

Comment:
The tests is supposed to update an existing instance, overriding `id` is not a correct thing to do in such case.

Looking at the original test again, the `instance.id` in the before-save hook should not match `record.id`, because the `record.id` is not know at the time when the hook is triggered. Instead, the id is a new unique value that is discarded later.

IMO the test should assert the last generated id value, which is pretty easy to get via `'' + lastId`. We should also check that the `record` returned by `findOrCreate` has the correct (existing) id.

```js
TestModel.observe('before save', pushContextAndNext());

TestModel.findOrCreate(
  { where: { name: existingInstance.name } },
  { name: existingInstance.name },
  function(err, record, created) {
    if (err) return done(err);
    created.id.should.eql(existingInstance.id);
    observedContexts.should.eql(aTestModelCtx({ instance: {
      id: getLastGeneratedUid(),
      name: existingInstance.name,
      extra: undefined
    }}));
    done();
  });

// at the bottom of the file, close to `function uid()`:
function getLastGeneratedUid() {
  return '' + lastId;
}
```

Thoughts?